src/compiler/moduleNameResolver.ts

```diff
@@ -1340,45 +1350,51 @@ function nodeNextModuleNameResolver(moduleName: string, containingFile: string,
     if (compilerOptions.resolveJsonModule) {
-        extensions = [...extensions, Extensions.Json];
+        extensions |= Extensions.Json;
     }
-    return nodeModuleNameResolverWorker(features | esmMode, moduleName, containingDirectory, compilerOptions, host, cache, extensions, redirectedReference);
+    return nodeModuleNameResolverWorker(features | esmMode, moduleName, containingDirectory, compilerOptions, host, cache, extensions, /*isConfigLookup*/ false, redirectedReference);
 }
 
 function tryResolveJSModuleWorker(moduleName: string, initialDir: string, host: ModuleResolutionHost): ResolvedModuleWithFailedLookupLocations {
-    return nodeModuleNameResolverWorker(NodeResolutionFeatures.None, moduleName, initialDir, { moduleResolution: ModuleResolutionKind.NodeJs, allowJs: true }, host, /*cache*/ undefined, jsOnlyExtensions, /*redirectedReferences*/ undefined);
+    return nodeModuleNameResolverWorker(
+        NodeResolutionFeatures.None,
+        moduleName,
+        initialDir,
+        { moduleResolution: ModuleResolutionKind.NodeJs, allowJs: true },
+        host,
+        /*cache*/ undefined,
+        Extensions.JavaScript,
+        /*isConfigLookup*/ false,
+        /*redirectedReferences*/ undefined);
 }
 
 export function nodeModuleNameResolver(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost, cache?: ModuleResolutionCache, redirectedReference?: ResolvedProjectReference): ResolvedModuleWithFailedLookupLocations;
 /** @internal */ export function nodeModuleNameResolver(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost, cache?: ModuleResolutionCache, redirectedReference?: ResolvedProjectReference, lookupConfig?: boolean): ResolvedModuleWithFailedLookupLocations; // eslint-disable-line @typescript-eslint/unified-signatures
-export function nodeModuleNameResolver(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost, cache?: ModuleResolutionCache, redirectedReference?: ResolvedProjectReference, lookupConfig?: boolean): ResolvedModuleWithFailedLookupLocations {
+export function nodeModuleNameResolver(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost, cache?: ModuleResolutionCache, redirectedReference?: ResolvedProjectReference, isConfigLookup?: boolean): ResolvedModuleWithFailedLookupLocations {
     let extensions;
-    if (lookupConfig) {
-        extensions = tsconfigExtensions;
+    if (isConfigLookup) {
+        extensions = Extensions.Json;
     }
     else if (compilerOptions.noDtsResolution) {
-        extensions = [Extensions.TsOnly];
-        if (compilerOptions.allowJs) extensions.push(Extensions.JavaScript);
-        if (compilerOptions.resolveJsonModule) extensions.push(Extensions.Json);
+        extensions = Extensions.ImplementationFiles;

```

Comment:
this is the part where we are saying always looks for js and report error if allowJs is not on correct.

Reply:
This is actually only used by Find Source Definition, which always passes `allowJs`. But I removed the condition to be consistent with other resolution modes which always resolve JS, yes.

Reply:
And to be clear, normal resolution has resolved JS and reported errors afterwards for like 6 years; that’s not new here 😄